stellarobservatory/intactness.py

```diff
@@ -0,0 +1,34 @@
+"""Algorithm for determining B-intact nodes given a set B of nodes."""
+from typing import Tuple, Callable, Set, Type
+
+from stellarobservatory.quorum_intersection import quorum_intersection
+from stellarobservatory.quorums import greatest_quorum
+
+
+def intact_nodes(fbas: Tuple[Callable[[Set[Type], Type, Set[Type]], bool], Set[Type]],
+                 b_nodes: set):
+    """
+    Takes an FBAS F (having quorum intersection) with set of nodes V and B ⊆ V and
+    returns the set of all B-intact nodes.
+    """
+    is_slice_contained, all_nodes = fbas
+    current = all_nodes.difference(b_nodes)
+    while True:
+        greatest_q = greatest_quorum(is_slice_contained, current, set(), set())
+        cur_fbas = (is_slice_contained, all_nodes.difference(greatest_q))
+        # determine whether F^{V\Q} has quorum intersection:
+        result = quorum_intersection(cur_fbas, greatest_q)
```

Comment:
It's a bit annoying that one has to pass in `D` into `quorum_intersection` *and* modify the nodes in the `fbas` param passed in as well (to compute `quorumIntersection(F^{V\D})`).

Reply:
Why don't we just set the `cur_fbas` to the actual modified FBAS à la `cur_fbas = (cur_is_slice_contained, all_nodes.difference(greatest_q))` where `cur_is_slice_contained` is defined in the loop as:
```python
def cur_is_slice_contained(candidate, node):
    if node in greatest_q:
       raise ValueError('node not defined in FBAS')
    return is_slice_contained(candidate.union(greatest_q), node)
```
This way the nodes in `greatest_q` just don't matter anymore, right? Can you verify @TorstenStueber or @Liamsi?

Reply:
This looks like an elegant way to get rid of the additional parameter (`without_d`) everywhere and still have the same functionality. Let me verify this and change this PR accordingly. I think this should work and we can have the "old" version of `greatest_quorum` without a changed signature and a param that is rarely used anywhere (namely only in `intact_nodes`).

Reply:
Hmm, we want the method (`cur_is_slice_contained`) to return true iff There is a v in V\Q and s in S(v): s\Q ⊆U (in the "FBAS without Q"). Looking at the suggestion it seems we instead return true if: there is a v in V\Q (asserted via the error) and an s in S(v):  s ⊆ U∪Q. I wrote Q instead greatest_q here. How are these equivalent?

Reply:
All tests pass with the suggestion though (only pylint &B mypy need some love): https://github.com/andrenarchy/stellar-observatory/commit/e83363491ad07ff97e44304241c32b15735b5ecc
We need more tests ...